### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by hamidjolany**
## Link to Click Up ticket

[Click Up Reference](https://app.clickup.com/t/1cdjvrt)

### *What is the problem we're trying to solve*

Account API Implementation and unit testing

### *What changes have we introduced*

the account entity was made based on the latest specification mentioned in bellow
[Specification reference](https://docs.google.com/document/d/11MpBHs0NFn14dYQXehIXOZt_RpOydoC5LTHva2OSLCc/edit)

#### _Checklist_

**Endpoint mentioned in bellow**

- [ ] GET /accounts?targetId={targetId}&type={accountType}
- [ ] GET /accounts/arrears?type={accountType}&sort={propertyName}&direction={asc/desc}
- [ ] GET /accounts/{id}
- [ ] POST /accounts
- [ ] PATCH /accounts/{id}

**Unit testing**

- [ ] Boundary request and response
- [ ] Controllers
- [ ] Domains
- [ ] Factories
- [ ] Gateways
- [ ] Use cases
- [ ] Infrastructures

**Integration testing**

- [ ] Gateways
- [ ] Use cases

### CODE REVIEW

File: AccountsApi.Tests/V1/Factories/EntityFactoryTest.cs [hunk 1]
```diff
@@ -1,39 +1,211 @@
...
-        // any edge cases that might exist.
-        [Test]
+        [Fact]
         public void CanMapADatabaseEntityToADomainObject()
         {
-            var databaseEntity = _fixture.Create<AccountDbEntity>();
-            var entity = databaseEntity.ToDomain();
+            var entity = new AccountDbEntity
+            {
+                Id = Guid.Parse("82aa6932-e98d-41a1-a4d4-2b44135554f8"),
```
Comment (LBHSPreston) [inline AccountsApi.Tests/V1/Factories/EntityFactoryTest.cs:20]: These tests need to be a bit more dynamic, either using trangulation or dynamically generated fake data

File: AccountsApi.Tests/V1/Infrastructure/DynamoDbInitilisationExtensionsTests.cs [hunk 2]
```diff
@@ -1,20 +1,21 @@
...
 namespace AccountsApi.Tests.V1.Infrastructure
 {
-    [TestFixture]
+    [Collection("MainCollection")]
     public class DynamoDbInitilisationExtensionsTests
     {
-        [TestCase(null)]
-        [TestCase("false")]
-        [TestCase("true")]
+        [Theory]
```
Comment (LBHSPreston) [inline AccountsApi.Tests/V1/Infrastructure/DynamoDbInitilisationExtensionsTests.cs:15]: This is a good example of triangulated tests as per my previous comment.  Maybe a bit more of this.

File: AccountsApi/Startup.cs [hunk 3]
```diff
@@ -36,14 +37,15 @@ public Startup(IConfiguration configuration)
 
         public IConfiguration Configuration { get; }
         private static List<ApiVersionDescription> _apiVersions { get; set; }
-        //TODO update the below to the name of your API
-        private const string ApiName = "Your API Name";
+
+        private const string ApiName = "AccountAPI";
```
Comment (LBHSPreston) [inline AccountsApi/Startup.cs:41]: This should be pluralised as per our namespace?

File: AccountsApi/V1/Controllers/AccountApiController.cs [hunk 4]
```diff
@@ -1,68 +1,198 @@
...
+                return BadRequest(new BaseErrorResponse((int) HttpStatusCode.BadRequest, ModelState.GetErrorMessages()));
+            }
+
+            var result = await _updateUseCase.ExecuteAsync(accountResponseObject).ConfigureAwait(false);
+
+            return Ok(result);
+        }
+
+        /// <summary>
+        /// Get a list of Arrear Account models
```
Comment (LBHSPreston) [inline AccountsApi/V1/Controllers/AccountApiController.cs:170]: Should be arrears instead of arrear.

File: AccountsApi/V1/Domain/ConsolidatedCharges.cs [hunk 5]
```diff
@@ -1,14 +1,29 @@
 using System;
 using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
 using System.Linq;
 using System.Threading.Tasks;
 
 namespace AccountsApi.V1.Domain
 {
-    public class ConsolidatedCharges
+    public class ConsolidatedCharge
```
Comment (LBHSPreston) [inline AccountsApi/V1/Domain/ConsolidatedCharges.cs:9]: Should be ConsolidatedCharges.  There would probably not be a need to consolidate if there is only one charge.

File: AccountsApi/V1/Domain/Enum.cs [hunk 6]
```diff
@@ -1,20 +1,44 @@
...
 using System.Text.Json.Serialization;
-using System.Threading.Tasks;
 
 namespace AccountsApi.V1.Domain
 {
     [JsonConverter(typeof(JsonStringEnumConverter))]
     public enum TargetType
     {
-        Housing, Garage
+        Estate, Block, Core, Flat
```
Comment (LBHSPreston) [inline AccountsApi/V1/Domain/Enum.cs:8]: We may need to revisit this.  Do we know if these hardcoded enums will ever change?  This might have to be  a configurable item or a reference lookup.

File: AccountsApi/V1/Domain/Enum.cs [hunk 7]
```diff
@@ -1,20 +1,44 @@
...
     }
 
     [JsonConverter(typeof(JsonStringEnumConverter))]
     public enum AccountStatus
     {
         Active, Suspended, Ended
```
Comment (LBHSPreston) [inline AccountsApi/V1/Domain/Enum.cs:14]: We may need to revisit this.  Do we know if these hardcoded enums will ever change?  This might have to be  a configurable item or a reference lookup.

File: AccountsApi/V1/Domain/Enum.cs [hunk 8]
```diff
@@ -1,20 +1,44 @@
...
     }
+
+    [JsonConverter(typeof(JsonStringEnumConverter))]
+    public enum Title
```
Comment (LBHSPreston) [inline AccountsApi/V1/Domain/Enum.cs:18]: We may need to revisit this.  Do we know if these hardcoded enums will ever change?  This might have to be  a configurable item or a reference lookup.

File: AccountsApi/V1/Domain/Enum.cs [hunk 9]
```diff
@@ -1,20 +1,44 @@
...
+    {
+        Mr,
+        Mrs,
+        Ms,
+        Dr
+    }
+
+    [JsonConverter(typeof(JsonStringEnumConverter))]
+    public enum AccountType
```
Comment (LBHSPreston) [inline AccountsApi/V1/Domain/Enum.cs:27]: We may need to revisit this.  Do we know if these hardcoded enums will ever change?  This might have to be  a configurable item or a reference lookup.

File: AccountsApi/V1/Domain/Enum.cs [hunk 10]
```diff
@@ -1,20 +1,44 @@
...
+    {
+        Master, Recharge, Sundry
+    }
+
+    [JsonConverter(typeof(JsonStringEnumConverter))]
+    public enum RentGroupType
```
Comment (LBHSPreston) [inline AccountsApi/V1/Domain/Enum.cs:33]: We may need to revisit this.  Do we know if these hardcoded enums will ever change?  This might have to be  a configurable item or a reference lookup.  There is a reference lookups API we can probably reuse for this.

File: AccountsApi/V1/Factories/EntityFactory.cs [hunk 11]
```diff
@@ -7,37 +9,89 @@ public static class EntityFactory
...
+                CreatedBy = account.CreatedBy,
+                CreatedDate = account.CreatedDate,
+                LastUpdatedBy = account.LastUpdatedBy,
+                LastUpdatedDate = account.LastUpdatedDate,
+                StartDate = account.StartDate,
+                TargetId = account.TargetId,
+                TargetType = account.TargetType,
+                AccountType = account.AccountType,
+                AgreementType = account.AgreementType,
+                RentGroupType = account.RentGroupType,
```
Comment (LBHSPreston) [inline AccountsApi/V1/Factories/EntityFactory.cs:92]: Possible null reference exception?  Might be a good idea to check for null for these complex objects from a different source.

File: AccountsApi/V1/Gateways/DynamoDbGateway.cs [hunk 12]
```diff
@@ -1,59 +1,60 @@
...
+        public async Task<List<Account>> GetAllAsync(Guid targetId, AccountType accountType)
         {
-            ScanCondition scanCondition_targetid = new ScanCondition("TargetId", Amazon.DynamoDBv2.DocumentModel.ScanOperator.Equal, targetId);
             List<ScanCondition> scanConditions = new List<ScanCondition>();
-            scanConditions.Add(scanCondition_targetid);
 
-            List<AccountDbEntity> data = await _dynamoDbContext
-                .ScanAsync<AccountDbEntity>(scanConditions)
-                .GetRemainingAsync()
+            ScanCondition scanConditionTargetId = new ScanCondition("TargetId", ScanOperator.Equal, targetId);
```
Comment (LBHSPreston) [inline AccountsApi/V1/Gateways/DynamoDbGateway.cs:50]: We need to evaluate the use of scans on our DynamoDb tables, particularly where we are filtering on more than one property or index, it is a very expensive operation and generally not good practice.  We can consider using query vs scan operations or look for an alternative way of querying our DynamoDb tables.

File: AccountsApi/V1/Gateways/DynamoDbGateway.cs [hunk 13]
```diff
@@ -1,59 +1,60 @@
...
-                await AddAsync(account).ConfigureAwait(false);
-            }
+            ScanCondition scanConditionAccountType = new ScanCondition("AccountType", ScanOperator.Equal, accountType);
+            ScanCondition scanConditionAccountBalance = new ScanCondition("AccountBalance", ScanOperator.LessThan, 0.0M);
+
+            scanConditions.Add(scanConditionAccountType);
+            scanConditions.Add(scanConditionAccountBalance);
+
+            var data = await _wrapper
+                .ScanAsync(_dynamoDbContext, scanConditions)
```
Comment (LBHSPreston) [inline AccountsApi/V1/Gateways/DynamoDbGateway.cs:40]: We need to evaluate the use of scans on our DynamoDb tables, particularly where we are filtering on more than one property or index, it is a very expensive operation and generally not good practice. We can consider using query vs scan operations or look for an alternative way of querying our DynamoDb tables.

**Review by LBHSPreston — CHANGES_REQUESTED** [review]
Quite a hefty pull request.  Moving forward we really need to break these down into more manageable chunks.

File: AccountsApi/V1/Gateways/DynamoDbGateway.cs (hunk 12 above)
Comment (hamidjolany) [reply AccountsApi/V1/Gateways/DynamoDbGateway.cs]: should we define two global index on these field?

**Review by LBHSPreston — CHANGES_REQUESTED** [review]
Please complete the details of the pull request comment so it is clear what this pull request covers.  Please include information such as reason for the change, link to any task tickets (JIRA, Clickup, Trello, etc).  It appears that this PR covers a number of tasks.  Moving forward pull requests must be for a single task/ticket unless we are performing a release.
